feat: handle unzip errors

Jan200101 opened the pull request:
fixes #238
fixes #240

handles unzip failures gracefully with visual feedback in the UI

![image](https://github.com/user-attachments/assets/f071d8f7-9cfc-4290-9f19-d485933a8739)

0neGal commented:
As it stands, I'm not sure I'd agree that this fixes those issues, but were you to add something like "this may happens due to file permissions" or similar, it'd be a different story. But even so, those issues should be caught much earlier than the unzipping process imo

Jan200101 commented:
sure, but this should be done regardless.
this is just a blanket solution that gracefully deals with *all* errors during unzipping, actually checking if we have the permission to put files in place could be done ahead of time though we should look to avoid what FlightCore does and check for a hardcoded path

0neGal commented:
> sure, but this should be done regardless.
> this is just a blanket solution that gracefully deals with _all_ errors during unzipping

Totally agree

> actually checking if we have the permission to put files in place could be done ahead of time though we should look to avoid what FlightCore does and not check for a hardcoded path

Also agree, however keep in mind, [there was already some attempt to make this be a thing](https://github.com/0neGal/viper/blob/main/src/modules/gamepath.js#L83), and it did work in my testing on Linux, as mentioned in those issues, I've however not been able to get around to if it's just on Windows it breaks, and why...

Jan200101 commented:
> Also agree, however keep in mind, [there was already some attempt to make this be a thing](https://github.com/0neGal/viper/blob/main/src/modules/gamepath.js#L83), and it did work in my testing on Linux, as mentioned in those issues, I've however not been able to get around to if it's just on Windows it breaks, and why...

Oh yeah thats kinda expected.
`fs.accessSync` doesn't really work with directories on windows, libuv does some minor checks and then pretends everything is a-okay since access isn't controlled by metadata but by a lot more complicated stuff that cannot be sanely put into the RWX format.

what could be added after the accessSync is an attempt to create a file and delete it.

0neGal commented:
I figured Windows did some dumb stuff underneath or handled something different oh well, any fix is appreciated really... Whether it's recursively checking the files inside, or creating a file etc

Jan200101 commented:
permission check for Windows was added.
I didn't make it check specifically for windows since this is such a minor thing.
![image](https://github.com/user-attachments/assets/5fabe0db-2166-4288-b90d-e3922d27cd00)

0neGal commented on src/modules/gamepath.js:
It might be worth simply changing the `(folder)` argument to `(folder = settings.gamepath())` instead of doing `folder || settings.gamepath()` everywhere...

Jan200101 replied:
should I also do this for `gamepath.exists` for the sake of keeping it the same?

Jan200101 replied:
also, would this even be a good idea to assign it as the default argument?
I'm not too familiar with how its handled in JS but in other languages its evaluated ahead of time and may have the wrong value.
might be better to just put `folder = folder || settings().gamepath` in the function

0neGal replied:
> should I also do this for `gamepath.exists` for the sake of keeping it the same?

Nah it wouldn't really make sense to do when it's only used once.

> also, would this even be a good idea to assign it as the default argument? I'm not too familiar with how its handled in JS but in other languages its evaluated ahead of time and may have the wrong value. might be better to just put `folder = folder || settings().gamepath` in the function

It gets evaluated when the function is run, which is fine with a function like `settings()`, you'd be shaving off microseconds by doing `folder = folder || settings().gamepath`, at least it would probably be something like that.

Either is fine, my main gripe was really just doing `folder || settings().gamepath` multiple times :p

Jan200101 commented:
the last thing I want to propose here is making `gamepath-lost-perms` create a toast message that would call `setpath` on click.

Currently it creates an alert box and opens the setpath dialog on loop until you select a path with permissions or until the user terminates the app because `gamepath.setting` keeps being set to false when the file dialog is closed.

0neGal commented:
The original intention was that the dialog would show, and it'd disable the buttons beside the gamepath related ones, letting you mount the drive or fix permissions etc without having to set the gamepath again. Just to clarify, are you saying the permission alert continuously show keeps popping up? Because that would be unintentional

Jan200101 commented:
> Just to clarify, are you saying the permission alert continuously show keeps popping up? Because that would be unintentional

yes, that is what is happening.

The loop goes like this:
setInterval(gamepath.js:L193) emits `gamepath-lost-perms`
`gamepath-lost-perms` emits `setpath` if `gamepath.setting` is false and sets it to true
`setpath` calls `gamepath.set`
`gamepath.set` opens the file dialog via `dialog.showOpenDialog` and sets `gamepath.setting` to `false` in its callback handler.
because `gamepath.setting` is false `gamepath-lost-perms` can start the dialog again.

0neGal commented:
I see, I suppose you could do:
```js
ipcMain.on("gamepath-lost-perms", async (e, selected_gamepath) => {
	if (! gamepath.setting && ! gamepath.lost_perms == selected_gamepath) {
		gamepath.lost_perms = selected_gamepath;

		// remaining code ...
	}
})
```
And then clear `gamepath.lost_perms` whenever a gamepath is set or the permissions have been restored.

Jan200101 commented on src/modules/gamepath.js:
```diff
@@ -163,13 +168,15 @@ gamepath.set = async (win, force_dialog) => {
 
 			if (! gamepath.has_perms(res.filePaths[0])) {
-				ipcMain.emit("missing-perms", null, res.filePaths[0]);
+				// the global handler will deal witih this
```
I removed this here since it doesn't actually work if `gamepath.settings` is true.
since there is already a global loop that will check for it, its best left for it to handle.

0neGal replied:
Keep in mind, that also prevents the `gui.gamepath.missing_perms` alert from showing, as the handler for `missing-perms` shows that.

Jan200101 replied:
oddly enough, I cannot reproduce the double alert when amending the commit.
possible race condition? will remove it for now.

0neGal replied:
Double alerts shouldn't technically speaking be possible, given `alert()` is synchronous and code blocking, unless you mean that another opens right after the other, I presume the 2 duplicate alerts had the exact same text?

Jan200101 replied:
you just tipped me off to what it actually is:
the alerts are synchronous but setTimeout runs separately
it finds that it lacks perms and calls `gamepath-lost-perms` while `missing-perms` is still showing an alert.
Because of this `gamepath.setting` is false and thus it queues up another alert.

adding `gamepath.setting = true` to both the `found-missing-perms` and `missing-perms` handlers should fix this

0neGal replied:
I suppose that would work, albeit it's not actually setting the gamepath, so the variable name wouldn't actually make sense anymore, but besides that, this would fix it.

Jan200101 replied:
yeah, but both call `setpath` directly after the alert which will handle `gamepath.settings`
its really just a way of making the global loop not start more dialogs routines

0neGal replied:
Fair, I think this is a fix then.

0neGal commented on src/modules/gamepath.js:
```diff
@@ -55,8 +55,9 @@ ipcMain.on("missing-perms", async (e, selected_gamepath) => {
-	if (! gamepath.setting) {
-		gamepath.setting = true;
+	if (! gamepath.setting && gamepath.lost_perms != selected_gamepath) {
+		console.log(gamepath.lost_perms);
```
I presume this was a debug console log that made it into a commit?

Jan200101 replied:
correct, will amend

Jan200101 commented:
Anything still needed for this to be merged?
- general unzip failures are handles and propagated up with the ability to print specific messages for certain failures
    - a message for permission denied was added, though it may be best to omit that for now since the existing permission check *should* handle this. a follow-up PR could add it and other checks in. 
- permissions check was updated to check for permissions and try a filesystem operation for maximum coverage
- toast over IPC was fixed
- race condition for permission check was fixed

0neGal commented:
Besides localization, this should be able to be merged.

Jan200101 commented:
reverted unneeded localization changes, added german translation.
Poked Alystrasz on Discord for a French translation, would need to ask @XNovaDelta for the spanish one.
I don't know if KenMizz is still away, so I'll try finding someone for the chinese translation again.

Jan200101 commented:
was asked to ping here so they are aware when they are back @Alystrasz

Alystrasz commented on src/lang/fr.json:
```diff
@@ -43,6 +43,7 @@
 			"current": "Version actuelle :",
 			"download_done": "Téléchargement terminé ! Extraction des fichiers...",
 			"downloading": "Téléchargement en cours...",
+			"failed": "Mise à jour échoué !",
```
```suggestion
			"failed": "Échec de la mise à jour !",
```

Jan200101 commented:
as far as I am aware, only the spanish localization is missing.
